pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -373,11 +376,41 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 		return err
 	}
 	if !reflect.DeepEqual(node.Annotations, oldNode.Annotations) {
+		// retain old north interfaces annotation
+		var oldNorthInterfacesAnnotation networkv1.NorthInterfacesAnnotation
+		if ann, exists := oldNode.Annotations[networkv1.NorthInterfacesAnnotationKey]; exists {
+			oldNorthInterfacesAnnotation, err = networkv1.ParseNorthInterfacesAnnotation(ann)
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse north interfaces annotation for multi-networking", "nodeName", oldNode.Name)
+			}
+		}
+
 		if err = utilnode.PatchNodeMultiNetwork(ca.client, node); err != nil {
 			nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
 			klog.ErrorS(err, "Failed to update the node annotations and capacity for multi-networking", "nodeName", node.Name)
 			return err
 		}
+
+		// calculate updates to multinetwork node count metric based on new north interfaces annotation
+		if ann, exists := node.Annotations[networkv1.NorthInterfacesAnnotationKey]; exists {
+			newNorthInterfacesAnnotation, err := networkv1.ParseNorthInterfacesAnnotation(ann)
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse north interfaces annotation for multi-networking", "nodeName", node.Name)
+			}
+
+			multiNetworkNodeDiff := make(map[string]float64)
+			for _, ni := range oldNorthInterfacesAnnotation {
+				multiNetworkNodeDiff[ni.Network] -= 1
+			}
+			for _, ni := range newNorthInterfacesAnnotation {
+				multiNetworkNodeDiff[ni.Network] += 1
+			}
+			for network, diff := range multiNetworkNodeDiff {
+				if diff != 0 {
+					multiNetworkNodes.WithLabelValues(network).Add(diff)
+				}
+			}
```

Comment:
You can avoid three loops if you just decrement the metric in the first loop and increment it in the second loop. You don't need to calculate the diff.